Search strings

aartaka opened the pull request:
This adds search string parsing for Tab-separated paths, as per [RFC4266](https://www.rfc-editor.org/rfc/rfc4266.html#section-2.2). It allows constructing search lines with search terms derived from URIs, which wasn't possible before this patch.

There's also parsing of Gopher+ path part, but it's not used anywhere. It doesn't hurt having this spec-conformant element, though.

aartaka committed:
cl-gopher: Parse search strings and Gopher+ strings per RFC4266.

This adds search string parsing for Tab-separated paths, as per
RFC4266. It allows constructing search lines with search terms from
URIs, which wasn't possible before this patch.

There's also parsing of Gopher+ path part, but it's not used
anywhere. It doesn't hurt having this spec-conformant element, though.

aartaka commented:
I'd also add proper term-printing, so that the parsed and the printed URIs are consistent, but that can wait. #1 is a good step towards this, and once it's merged, adding term printing should be easy.

knusbaum commented on cl-gopher.lisp:
```diff
@@ -478,16 +478,23 @@
+             :selector selector
+             :hostname host
+             :port port
+             (when (and terms (eq item-type :search-line))
+               `(:terms ,terms))))))
```
For consistency you should add a `terms` slot to the `gopher-line` [class definition](https://github.com/knusbaum/cl-gopher/blob/29fbe52e3f377e24e8bde77b00789a1e32a75b90/cl-gopher.lisp#L72-L75)

aartaka replied:
Done!

knusbaum commented on cl-gopher.lisp:
```diff
@@ -478,16 +478,23 @@
+           (tab-split-selector (uiop:split-string selector :separator '(#\Tab)))
+           (selector (first tab-split-selector))
+           (terms (second tab-split-selector))
+           (gopher+string (third tab-split-selector))
```
Why bother binding `gopher+string` if we're not going to do anything with it?

aartaka replied:
That was one of the questions I had about this patch. I'll remove it then!

knusbaum commented on cl-gopher.lisp:
```diff
@@ -478,16 +478,23 @@
+           (path (quri:url-decode (quri:uri-path uri)))
            (item-type (compute-item-type uri path))
            (selector (compute-selector uri path))
+           (tab-split-selector (uiop:split-string selector :separator '(#\Tab)))
```
Can we use the standard `split-sequence` rather than relying on `uiop`?

aartaka replied:
But wait, there's no `split-sequence` in CLHS 0_o

UIOP is a de-facto standard and all the implementations package it. I see no problem with using it there :) I mean, I can rewrite it using `subseq` and friends, but why not use something that makes it better and is provided by all the modern implementations?

knusbaum replied:
Sorry, by "standard" I meant "what I'm already using in `cl-gopher`".

I have a `split-sequence` dependency: https://github.com/knusbaum/cl-gopher/blob/master/cl-gopher.asd#L6
Which I use in `read-gopher-line`: https://github.com/knusbaum/cl-gopher/blob/master/cl-gopher.lisp#L270

I would prefer to either use that `split-sequence` or replace its use with `uiop:split-string` and remove the dependency to be consistent.

knusbaum reviewed (changes requested):
Thanks for the patch! I have a couple changes I'd like before merging.

aartaka committed:
cl-gopher: Remove unused gopher+string.

aartaka committed:
cl-gopher: Move terms slot to gopher-line.

aartaka commented:
@knusbaum, all addressed except the UIOP replacement, as I don't see any problem in using UIOP :)

aartaka committed:
cl-gopher: Use uiop:split-string instead of split-sequence.

aartaka committed:
Remove split-sequence dependency.

aartaka committed:
cl-gopher: Print search terms to URI, if present.

aartaka commented:
@knusbaum, I've removed `split-sequence` in favor of UIOP, rebased over master and added term printing to `uri-for-gopher-line`. It's merge-ready, I believe.

knusbaum reviewed (approved):
@aartaka Looks awesome, thanks!

aartaka commented:
Thanks for this library :)
